controllers/teleport_master_controller/teleport_master_controller.cpp

```diff
@@ -0,0 +1,169 @@
+    // of the robot supervisor. Other arguments are def arguments of robots that should be controlled by
+    // this supervisor
+    if (argc < 2) {
+        std::cerr << "Usage: " << argv[0] << " <DEF>" << std::endl;
+        return EXIT_FAILURE;
+    }
+    // Load self-def argument which will be used to identify the robot using the webots getFromDef function
+    std::string def = argv[1];
+    
+    // Load def arguments of other robots in the field into a vector
+    std::vector<webots::Node*> otherRobotsNodes;
+    // Only one supervisor instance can exist in a single controller. As such the same one will be used
+    // every time
+    webots::Supervisor supervisor = webots::Supervisor();
+    for (int i = 2; i < argc; i++)
+      {
+        // Load the Node of the robot by def argument and add to vector
+        webots::Node* tempNode = supervisor.getFromDef(argv[i]);
+        otherRobotsNodes.push_back(tempNode);
+      }
+    
+    // Load config file
+    try {
+        YAML::Node config = YAML::LoadFile("config.yaml");
+        rotations         = config["rotations"].as<std::vector<std::array<double, 4>>>();
+    }
+    catch (const YAML::BadFile& e) {
+        std::cerr << e.msg << std::endl;
+        return 1;
+    }
+    catch (const YAML::ParserException& e) {
+        std::cerr << e.msg << std::endl;
+        return 2;
+    }
+
+    // Open the existing log file and immediately close it to overwrite it for a new instance of
+    // the program
+    std::ofstream log;
+    log.open("teleport_controller.log");
+    log.close();
+    
+    // Get the time step of the current world.
+    int timeStep = int(supervisor.getBasicTimeStep());
+    
+    // Generate random seed
+    std::random_device rd;   // Will be used to obtain a seed for the random number engine
+    std::mt19937 gen(rd());  // Standard mersenne_twister_engine seeded with rd()
+    
+    // Generate distributions for random number systems in the following loop
+    std::uniform_int_distribution<> xDistrib(0, 860);
+    std::uniform_int_distribution<> yDistrib(0, 580);
+    std::uniform_int_distribution<> rotDistrib(0, rotations.size() - 1);
+
+    while (supervisor.step(timeStep) != -1) {
+        
+        // Declare a vector of positions that will be saved as they are randomly generated, to be later
+        // applied to each robot
+        std::vector<std::array<double, 3>> positions = std::vector<std::array<double, 3>>();
+        // Loop through every robot
+        for (long unsigned int i = 0; i < otherRobotsNodes.size(); i++)
+        {
+          // Assume there is no collision
+          bool collision = false;
+          // newPos will be a "Proposed location" for a robot to teleport to
+          std::array<double, 3> newPos;
+          do
+          {
+            collision = false;
+            // Generate a new random location
+            newPos[0] = 4.3 - xDistrib(gen) / 100;
```

Comment:
Would be good to get the field hard coded values from the actual field if possible.

Reply:
I looked and looked around for this and had no luck, if you have any ideas that would be great but this was the only way I could figure out how to do it 😢 I can grab it from config too if that's cleaner but I couldn't work out how to do it dynamically

Reply:
I've just done this, what you need to do is change the RobocupSoccerField node in the world file to have a def, ie
```
DEF FIELD RobocupSoccerField {
  size "kid"
}
```
Then you can get it using the supervisor from the def.
```
webots::Node* field = supervisor.getFromDef("FIELD");
```
And, for example, getting the name of the field and printing it out:
```
std::cout << field->getField("size")->getSFString() << std::endl;
```
There isn't a specific field for these values so you would need to change the proto so you can grab them.

Reply:
Yeah okay, but you want it in the proto rather than the controller?
The thing is the range of random numbers I'm using isn't actually the size of the entire field - it's just the space inside the playing lines (Was getting weird errors when it was teleporting into goals by accident) is it still okay to have that in the proto?

Reply:
I think it's ok to put in a field for it in the proto, just to expose it so we can grab the value. But then we'd set the actual value inside the proto (not in the field part, but in the node tree) using the dimensions (I believe thats what A,B,C etc are) to do some Lua code to add whatever is needed up.
Might be better too to use the full field but add the goals and ball in to the collision detection.

Reply:
I can add the goals into the collision detection. I was told yesterday it probably won't be necessary but if you would like it I'll just need to grab them from def as well that's easy enough

Reply:
Okay so it appears that grabbing the goals to avoid collisions is going to be a major challenge... Because they're part of the "RobocupSoccerField" Node rather than objects in their own right, I am uncertain how to go about checking the actual goals. I can check for a collision but only if it has already moved there, which would still slow the simulation down.

Reply:
If it works as is currently then I don't think we should worry about it too much. You'd need to deal with transformations to do it as a child of the field, and it's probably fine as is.